### PR TITLE
Executable examples

### PR DESCRIPTION
**Pull request by RonghaoCui**
## Proposed changes

This PR aims to address #1629. We add main functions for examples and make sure examples can be executable using the minimal set of parameters. Some of examples have been added with the plot functionality. 

However, we found the convergence issue in benchmark 2d cases. This is solved temporarily by increasing convergence tolerance, as indicated in benchmark 2d examples. A new issue regarding this convergence issue as been created as #1657. 

## Types of changes

What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply._

- [x] Minor change (e.g., dependency bumps, broken links).
- [x] Bugfix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Testing (contribution related to testing of existing or new functionality).
- [ ] Documentation (contribution related to adding, improving, or fixing documentation).
- [ ] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [x] The documentation is up-to-date.
- [ ] Static typing is included in the update.
- [x] This PR does not duplicate existing functionality.
- [ ] The update is covered by the test suite (including tests added in the PR).
- [ ] If new skipped tests have been introduced in this PR, `pytest` was run with the `--run-skipped` flag.

### CODE REVIEW

File: src/porepy/applications/md_grids/gmsh_file_library/benchmark_2d_case_4/fracture_network_benchmark_2d_case_4.csv
Comment (IvarStefansson) [inline src/porepy/applications/md_grids/gmsh_file_library/benchmark_2d_case_4/fracture_network_benchmark_2d_case_4.csv:1]: Please explain why this file was changed.

File: src/porepy/examples/flow_benchmark_3d_case_2.py [hunk 1]
```diff
@@ -251,3 +251,64 @@ class FlowBenchmark3dCase2Model(  # type:ignore[misc]
...
+            },
+            "refinement_level": 0,  # 0, 1, or 2 for different mesh refinement levels
+        }
+
+        model = FlowBenchmark3dCase2Model(params)  # type: ignore[abstract]
+        pp.run_time_dependent_model(model, params)
+
+        sd_matrix = model.mdg.subdomains(dim=3)[0]
+        cc = sd_matrix.cell_centers
+        pressure = cast(
```
Comment (IvarStefansson) [inline src/porepy/examples/flow_benchmark_3d_case_2.py]: Is this plotting really necessary?

File: src/porepy/examples/flow_benchmark_3d_case_3.py [hunk 2]
```diff
@@ -129,3 +129,55 @@ class FlowBenchmark3dCase3Model(  # type:ignore[misc]
...
+    model = FlowBenchmark3dCase3Model(params)  # type: ignore[abstract]
+    pp.run_time_dependent_model(model, params)
+
+    sd_matrix = model.mdg.subdomains(dim=3)[0]
+    cc = sd_matrix.cell_centers
+    pressure = cast(
+        np.ndarray, model.equation_system.evaluate(model.pressure([sd_matrix]))
+    )
+
+    # Collect pressure solution along the vertical line of the domain.
```
Comment (IvarStefansson) [inline src/porepy/examples/flow_benchmark_3d_case_3.py]: Same comment as other 3d case. If we are to keep it, let's consider creating a plot solution over line function.

File: src/porepy/examples/fracture_damage.py [hunk 3]
```diff
@@ -486,3 +490,97 @@ class FractureDamageMomentumBalance(  # type: ignore[misc]
...
+            },
+            "time_manager": pp.TimeManager(np.arange(0, time_steps), 1, True),
+        }
+    )
+
+    return params_local
+
+
+# If executed as main, run simulation.
+if __name__ == "__main__":
```
Comment (IvarStefansson) [inline src/porepy/examples/fracture_damage.py]: This main statement is rather complex. Please add some documentation.

File: src/porepy/examples/mandel_biot.py [hunk 4]
```diff
@@ -1444,3 +1444,39 @@ class MandelModel(  # type: ignore[misc]
             - viscosity
 
     """
+
+
+# If executed as a script, run the verification
```
Comment (IvarStefansson) [inline src/porepy/examples/mandel_biot.py]: Here and throughout: Please add full stops at end of sentences.

File: src/porepy/examples/terzaghi_biot.py [hunk 5]
```diff
@@ -711,3 +711,21 @@ class TerzaghiModel(  # type: ignore[misc]
...
+            "fluid": pp.FluidComponent(**terzaghi_fluid_constants),  # type: ignore[arg-type]
+        },
+        "num_cells": 10,
+        "times_to_export": [],  # Suppress output for tests
+    }
+    model_biot = TerzaghiModel(model_params_poromech)  # type: ignore[abstract]
+    pp.run_time_dependent_model(model=model_biot)
+
+    p_biot = model_biot.results[0].approx_pressure
+    u_biot = model_biot.results[0].approx_consolidation_degree
```
Comment (IvarStefansson) [inline src/porepy/examples/terzaghi_biot.py]: What is the purpose of extracting the solutions here? Please motivate or remove.

**Review by IvarStefansson — COMMENTED** [review]
Looks good in the main. Please address my comments and request a new review.

File: src/porepy/examples/flow_benchmark_3d_case_2.py [hunk 6]
```diff
@@ -251,3 +251,64 @@ class FlowBenchmark3dCase2Model(  # type:ignore[misc]
...
+    for case_name, solid_constants in solid_cases.items():
+        params = {
+            "material_constants": {
+                "solid": solid_constants,
+            },
+            "refinement_level": 0,  # 0, 1, or 2 for different mesh refinement levels
+        }
+
+        model = FlowBenchmark3dCase2Model(params)  # type: ignore[abstract]
+        pp.run_time_dependent_model(model, params)
```
Comment (mikeljordan) [inline src/porepy/examples/flow_benchmark_3d_case_2.py]: The second argument should be a dict for solver parameters, not model parameters. If you are using default solver settings, you can remove `params` or split accordingly.

File: src/porepy/examples/flow_benchmark_3d_case_3.py [hunk 7]
```diff
@@ -129,3 +129,55 @@ class FlowBenchmark3dCase3Model(  # type:ignore[misc]
...
+    plt.figure(figsize=(6, 4))
+    params = {
+        "material_constants": {
+            "solid": solid_constants,
+        },
+        "refinement_level": 0,  # 0, 1, 2, or 3 for different mesh refinement levels
+    }
+
+    model = FlowBenchmark3dCase3Model(params)  # type: ignore[abstract]
+    pp.run_time_dependent_model(model, params)
```
Comment (mikeljordan) [inline src/porepy/examples/flow_benchmark_3d_case_3.py]: The same comment as above for the second argument applies here.

File: src/porepy/examples/fracture_damage.py [hunk 8]
```diff
@@ -486,3 +490,97 @@ class FractureDamageMomentumBalance(  # type: ignore[misc]
     Also contains specifics defining a test case in terms of the boundary conditions.
 
     """
+
+
+def common_params(time_steps: int, regime: str) -> dict:
```
Comment (mikeljordan) [inline src/porepy/examples/fracture_damage.py]: This function is only used by the __main__ block, so it should be inside it or at least guarded. As written, it becomes part of the module's public API.

File: src/porepy/examples/fracture_damage.py [hunk 9]
```diff
@@ -486,3 +490,97 @@ class FractureDamageMomentumBalance(  # type: ignore[misc]
...
+        }
+    )
+
+    return params_local
+
+
+# If executed as main, run simulation.
+if __name__ == "__main__":
+    # The model type can be "isotropic", "anisotropic", or "momentum_balance".
+    model_type = "momentum_balance"
```
Comment (mikeljordan) [inline src/porepy/examples/fracture_damage.py]: For a first-time user running this example, "isotropic" is the simplest and most intuitive default. The comment on line 538 is helpful, but the default should be the easiest case.

**Review by mikeljordan — COMMENTED** [review]
Nice work!, I left few comments.

File: src/porepy/applications/md_grids/gmsh_file_library/benchmark_2d_case_4/fracture_network_benchmark_2d_case_4.csv
Comment (RonghaoCui) [reply src/porepy/applications/md_grids/gmsh_file_library/benchmark_2d_case_4/fracture_network_benchmark_2d_case_4.csv:1]: Thank you for raising this question. This is because flow_benchmark_2d_case_4.py loads the fracture geometry through pp.applications.md_grids.fracture_sets.benchmark_2d_case_4(), which in turn calls pp.fracture_importer.network_from_csv(..., has_domain=False) on this file. 

Therefore, the CSV file for 2D fractures needs to follow the format expected by network_from_csv:
- first non-comment line = domain definition (xmin, ymin, xmax, ymax)
- remaining non-comment lines = fractures (start_x, start_y, end_x, end_y)

The previous CSV file with headings and fracture no. cannot match this required format, leading to a loading error. The CSV importer can load fractures without domain definition by setting has_domain=False. 

The update to this file was made to ensure that the input of benchmark 2d case 4 matches the format required by the current CSV importer. I also checked by search that the csv file, fracture_network_benchmark_2d_case_4.csv, is only used in pp.applications.md_grids.fracture_sets.benchmark_2d_case_4().

File: src/porepy/examples/flow_benchmark_3d_case_2.py (hunk 1 above)
Comment (RonghaoCui) [reply src/porepy/examples/flow_benchmark_3d_case_2.py]: To be honest, I think this plot is not quite necessary. I created this plot based on results of the corresponding paper for checking whether this 3D example can run well. We can remove them absolutely.

File: src/porepy/examples/flow_benchmark_3d_case_3.py (hunk 2 above)
Comment (RonghaoCui) [reply src/porepy/examples/flow_benchmark_3d_case_3.py]: I think we can remove them for sure. If we plan to keep it, we need to think about where to place the line function for extracting the solution for these two 3D cases. Do you have any suggestion on this?

File: src/porepy/examples/fracture_damage.py (hunk 9 above)
Comment (RonghaoCui) [reply src/porepy/examples/fracture_damage.py]: I agree. Now the "isotropic" type is set by default. This has also been mentioned in documentation.

File: src/porepy/examples/fracture_damage.py (hunk 3 above)
Comment (RonghaoCui) [reply src/porepy/examples/fracture_damage.py]: Done.

File: src/porepy/examples/fracture_damage.py (hunk 8 above)
Comment (RonghaoCui) [reply src/porepy/examples/fracture_damage.py]: Agree. This function has been moved into the main block.

File: src/porepy/examples/mandel_biot.py (hunk 4 above)
Comment (RonghaoCui) [reply src/porepy/examples/mandel_biot.py]: Done.

File: src/porepy/examples/terzaghi_biot.py (hunk 5 above)
Comment (RonghaoCui) [reply src/porepy/examples/terzaghi_biot.py]: I believe they were just used for testing output. I have removed them.

File: src/porepy/examples/flow_benchmark_3d_case_2.py (hunk 6 above)
Comment (RonghaoCui) [reply src/porepy/examples/flow_benchmark_3d_case_2.py]: Removed and increased the tolerance for the non-linear solver.

File: src/porepy/examples/flow_benchmark_3d_case_3.py (hunk 7 above)
Comment (RonghaoCui) [reply src/porepy/examples/flow_benchmark_3d_case_3.py]: Removed and increased the tolerance for the non-linear solver.

File: src/porepy/applications/md_grids/gmsh_file_library/benchmark_2d_case_4/fracture_network_benchmark_2d_case_4.csv
Comment (IvarStefansson) [reply src/porepy/applications/md_grids/gmsh_file_library/benchmark_2d_case_4/fracture_network_benchmark_2d_case_4.csv:1]: Thanks for the thorough explanation. In other words, you fixed something that was broken and which wasn't caught when we didn't execute this example. Well done!

File: src/porepy/examples/flow_benchmark_3d_case_3.py (hunk 2 above)
Comment (IvarStefansson) [reply src/porepy/examples/flow_benchmark_3d_case_3.py]: Looking more thoroughly at this, the simplifications made in this implementation mean that we should skip the plotting. We can consider reintroducing sampling and plotting along a line as a separate project later.

**Review by IvarStefansson — APPROVED** [review]

File: src/porepy/examples/flow_benchmark_3d_case_3.py (hunk 2 above)
Comment (RonghaoCui) [reply src/porepy/examples/flow_benchmark_3d_case_3.py]: Sure. Plots in 3d examples have been removed.
